app_check/integration_test/src/integration_test.cc

```diff
@@ -821,4 +919,191 @@ TEST_F(FirebaseAppCheckTest, TestFunctionsFailure) {
+  WaitForCompletion(
+      document.Set(firebase::firestore::MapFieldValue{
+          {"str", firebase::firestore::FieldValue::String("badfoo")},
+          {"int", firebase::firestore::FieldValue::Integer(456)}}),
+      "document.Set", firebase::firestore::kErrorPermissionDenied);
+  WaitForCompletion(document.Get(firebase::firestore::Source::kServer),
+                    "document.Get",
+                    firebase::firestore::kErrorPermissionDenied);
+
+  CleanupFirestore(firebase::firestore::kErrorPermissionDenied);
+}
+
+TEST_F(FirebaseAppCheckTest, TestFirestoreTransaction) {
+  InitializeAppCheckWithDebug();
+  InitializeApp();
+  InitializeFirestore();
+
+  firebase::firestore::DocumentReference document = CreateFirestoreDoc();
+
+  WaitForCompletion(
+      document.Set(firebase::firestore::MapFieldValue{
+          {"val", firebase::firestore::FieldValue::String("start")}}),
+      "document.Set 0");
+
+  std::vector<firebase::firestore::MapFieldValue> document_snapshots;
+  firebase::firestore::ListenerRegistration registration =
+      document.AddSnapshotListener(
+          [&](const firebase::firestore::DocumentSnapshot& result,
+              firebase::firestore::Error error_code,
+              const std::string& error_message) {
+            EXPECT_EQ(error_code, firebase::firestore::kErrorOk);
+            EXPECT_EQ(error_message, "");
+            document_snapshots.push_back(result.GetData());
+          });
+
+  WaitForCompletion(
+      document.Set(firebase::firestore::MapFieldValue{
+          {"val", firebase::firestore::FieldValue::String("update")}}),
+      "document.Set 1");
+
+  registration.Remove();
+  WaitForCompletion(
+      document.Set(firebase::firestore::MapFieldValue{
+          {"val", firebase::firestore::FieldValue::String("final")}}),
+      "document.Set 2");
+  EXPECT_THAT(
+      document_snapshots,
+      testing::ElementsAre(
+          firebase::firestore::MapFieldValue{
+              {"val", firebase::firestore::FieldValue::String("start")}},
+          firebase::firestore::MapFieldValue{
+              {"val", firebase::firestore::FieldValue::String("update")}}));
+}
+
+TEST_F(FirebaseAppCheckTest, TestFirestoreTransactionFailure) {
+  // Don't set up AppCheck
+  InitializeApp();
+  InitializeFirestore();
+
+  firebase::firestore::DocumentReference document = CreateFirestoreDoc();
+
+  bool received_pending_cache = false;
+  bool received_permission_denied = false;
+  firebase::firestore::ListenerRegistration registration =
+      document.AddSnapshotListener(
+          [&received_pending_cache, &received_permission_denied](
+              const firebase::firestore::DocumentSnapshot& result,
+              firebase::firestore::Error error_code,
+              const std::string& error_message) {
+            // We first expect one call for a successful change to the cache.
+            if (error_code == firebase::firestore::kErrorNone) {
```

Comment:
Asserting a very specific sequence of snapshots is brittle. Instead, I'd recommend just asserting that _eventually_ you get a "permission denied" error.

Reply:
Makes sense.  Changed it to just check that if it receives an Ok, the metadata is what is expected, but it no longer requires that an Ok happens.